File: web/src/types/public.ts
export interface EquityHistoryQuery {
  trader_id?: string | null
}

export interface EquityHistoryBatchRequest {
  trader_ids?: string[] | null
  hours?: number | null
}

export interface SymbolsQuery {
  exchange?: string | null
}

export interface KlinesQuery {
  symbol: string
  interval?: string | null
  limit?: number | null
  exchange?: string | null
}

export interface CryptoConfigPayload {
  transport_encryption: boolean
}

export interface CryptoPublicKeyPayload {
  transport_encryption: boolean
  public_key: string
}

export interface ServerIpPayload {
  public_ip: string
  message: string
}

export interface PublicCompetitionTraderPayload {
  trader_id: string
  trader_name: string
  ai_model: string
  exchange: string
  total_equity: number
  total_pnl: number
  total_pnl_pct: number
  position_count: number
  margin_used_pct: number
  is_running: boolean
}

export interface CompetitionListPayload {
  traders: PublicCompetitionTraderPayload[]
  count: number
}

export interface EquityHistoryPointPayload {
  timestamp: string
  total_equity: number
  available_balance: number
  total_pnl: number
  total_pnl_pct: number
  position_count: number
  margin_used_pct: number
  balance: number
}

export interface EquityHistoryBatchPayload {
  histories: Record<string, EquityHistoryPointPayload[]>
  errors: Record<string, string>
}

export interface PublicTraderConfigPayload {
  trader_id: string
  trader_name: string
  ai_model: string
  exchange_id: string
  strategy_id: string
  is_cross_margin: boolean
  show_in_competition: boolean
  scan_interval_minutes: number
  initial_balance: number
  is_running: boolean
  btc_eth_leverage: number
  altcoin_leverage: number
  trading_symbols: string
  custom_prompt: string
  override_base_prompt: boolean
  system_prompt_template: string
  use_ai500: boolean
  use_oi_top: boolean
}

export interface ExchangeSymbolPayload {
  symbol: string
  name: string
  category: string
}

export interface KlinePayload {
  openTime: number
  open: number
  high: number
  low: number
  close: number
  volume: number
  quoteVolume: number
  closeTime: number
}

export interface ExchangeSymbolsPayload {
  exchange: string
  symbols: ExchangeSymbolPayload[]
  count: number
}

export interface SupportedProviderTypePayload {
  providerType: string
  name: string
}

export interface SupportedExchangePayload {
  id: string
  name: string
  type: string
}

File: core/src/contracts/public.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::{
    clients::llm_chat::SupportedProviderType,
    clients::market_data::{MarketKline, MarketSymbol},
};

#[derive(Debug, Deserialize)]
pub struct EquityHistoryQuery {
    pub trader_id: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct EquityHistoryBatchRequest {
    pub trader_ids: Option<Vec<String>>,
    pub hours: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct SymbolsQuery {
    pub exchange: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct KlinesQuery {
    pub symbol: String,
    pub interval: Option<String>,
    pub limit: Option<i64>,
    pub exchange: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct CryptoConfigPayload {
    pub transport_encryption: bool,
}

#[derive(Debug, Clone, Serialize)]
pub struct CryptoPublicKeyPayload {
    pub transport_encryption: bool,
    pub public_key: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct PublicCompetitionTraderPayload {
    pub trader_id: String,
    pub trader_name: String,
    pub ai_model: String,
    pub exchange: String,
    pub total_equity: f64,
    pub total_pnl: f64,
    pub total_pnl_pct: f64,
    pub position_count: i64,
    pub margin_used_pct: f64,
    pub is_running: bool,
}

#[derive(Debug, Clone, Serialize)]
pub struct CompetitionListPayload {
    pub traders: Vec<PublicCompetitionTraderPayload>,
    pub count: usize,
}

#[derive(Debug, Clone, Serialize)]
pub struct EquityHistoryPointPayload {
    pub timestamp: String,
    pub total_equity: f64,
    pub available_balance: f64,
    pub total_pnl: f64,
    pub total_pnl_pct: f64,
    pub position_count: i64,
    pub margin_used_pct: f64,
    pub balance: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct EquityHistoryBatchPayload {
    pub histories: HashMap<String, Vec<EquityHistoryPointPayload>>,
    pub errors: HashMap<String, String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct PublicTraderConfigPayload {
    pub trader_id: String,
    pub trader_name: String,
    pub ai_model: String,
    pub exchange_id: String,
    pub strategy_id: String,
    pub is_cross_margin: bool,
    pub show_in_competition: bool,
    pub scan_interval_minutes: i32,
    pub initial_balance: f64,
    pub is_running: bool,
    pub btc_eth_leverage: i32,
    pub altcoin_leverage: i32,
    pub trading_symbols: String,
    pub custom_prompt: String,
    pub override_base_prompt: bool,
    pub system_prompt_template: String,
    pub use_ai500: bool,
    pub use_oi_top: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExchangeSymbolPayload {
    pub symbol: String,
    pub name: String,
    pub category: String,
}

impl From<MarketSymbol> for ExchangeSymbolPayload {
    fn from(value: MarketSymbol) -> Self {
        Self {
            symbol: value.symbol,
            name: value.name,
            category: value.category,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KlinePayload {
    #[serde(rename = "openTime")]
    pub open_time: i64,
    pub open: f64,
    pub high: f64,
    pub low: f64,
    pub close: f64,
    pub volume: f64,
    #[serde(rename = "quoteVolume")]
    pub quote_volume: f64,
    #[serde(rename = "closeTime")]
    pub close_time: i64,
}

impl From<MarketKline> for KlinePayload {
    fn from(value: MarketKline) -> Self {
        Self {
            open_time: value.open_time,
            open: value.open,
            high: value.high,
            low: value.low,
            close: value.close,
            volume: value.volume,
            quote_volume: value.quote_volume,
            close_time: value.close_time,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct ExchangeSymbolsPayload {
    pub exchange: String,
    pub symbols: Vec<ExchangeSymbolPayload>,
    pub count: usize,
}

#[derive(Debug, Clone, Serialize)]
pub struct SupportedProviderTypePayload {
    #[serde(rename = "providerType")]
    pub provider_type: String,
    pub name: String,
}

impl From<&SupportedProviderType> for SupportedProviderTypePayload {
    fn from(value: &SupportedProviderType) -> Self {
        Self {
            provider_type: value.provider_type.to_string(),
            name: value.name.to_string(),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct SupportedExchangePayload {
    pub id: String,
    pub name: String,
    #[serde(rename = "type")]
    pub exchange_kind: String,
}
